message_ix/tests/test_macro.py

```diff
@@ -7,145 +7,163 @@
-    assert "config" in c.data
-    assert "sector" in c.data["config"]
+def test_config(westeros_solved, w_data_path):
+    c = macro.prepare_computer(westeros_solved, data=w_data_path)
+    assert "config::macro" in c.graph
+    assert "sector" in c.get("config::macro")
+    assert {"Westeros"} == c.get("node::macro")
+    assert {"light"} == c.get("sector::macro")
 
-    # Removing a column from config and testing
-    data = c.data.copy()
-    data["config"] = c.data["config"][["node", "sector"]]
-    try:
-        macro.Calculate(s, data)
-    except KeyError as error:
-        assert 'Missing config data for "level"' in str(error)
+    # Missing columns in config raises an exception
+    data = c.get("data")
+    data["config"] = data["config"][["node", "sector"]]
+    c = macro.prepare_computer(westeros_solved, data=data)
+    with pytest.raises(Exception, match="level"):
+        c.get("check all")
 
-    # Removing config completely and testing
+    # Entirely missing config raises an exception
     data.pop("config")
-    try:
-        macro.Calculate(s, data)
-    except KeyError as error:
-        assert "Missing config in input data" in str(error)
+    c = macro.prepare_computer(westeros_solved, data=data)
+    with pytest.raises(Exception, match="config"):
+        c.get("check all")
 
-    c.read_data()
-    assert c.nodes == set(["Westeros"])
-    assert c.sectors == set(["light"])
 
+def test_calc_data_missing_par(westeros_solved, w_data):
+    data = w_data
 
-def test_calc_data_missing_par(westeros_solved):
-    s = westeros_solved
-    data = pd.read_excel(W_DATA_PATH, sheet_name=None, engine="openpyxl")
     data.pop("gdp_calibrate")
-    c = macro.Calculate(s, data)
-    pytest.raises(ValueError, c.read_data)
 
+    c = macro.prepare_computer(westeros_solved, data=data)
+    with pytest.raises(Exception, match="gdp_calibrate"):
+        c.get("check all")
+
+
+def test_calc_data_missing_ref(westeros_solved: Scenario, w_data):
+    """When "price_ref" is missing, the code computes it by extrapolation."""
+    data = w_data
+
+    # NB this doesn't work because extrapolating [511, 162, 161] backwards gives 1134, a
+    #    very high value; MACRO errors
+    # data.pop("cost_ref")
+
+    data.pop("price_ref")
+
+    c = macro.prepare_computer(westeros_solved, data=data)
+    c.get("check all")
+
+    westeros_solved.add_macro(data)
 
-def test_calc_data_missing_column(westeros_solved):
-    s = westeros_solved
-    data = pd.read_excel(W_DATA_PATH, sheet_name=None, engine="openpyxl")
-    # skip first data point
+
+def test_calc_data_missing_column(westeros_solved, w_data):
+    data = w_data
+
+    # Drop a column
     data["gdp_calibrate"] = data["gdp_calibrate"].drop("year", axis=1)
-    c = macro.Calculate(s, data)
-    pytest.raises(ValueError, c.read_data)
+
+    c = macro.prepare_computer(westeros_solved, data=data)
+    with pytest.raises(Exception, match="Missing expected columns.*year"):
```

Comment:
The formulation of `match` seems odd to me here, so just to make sure: is the raised error message really referencing something like `columns.gdp_calibrate.year`? The [MissingKeyError in genno](https://github.com/khaeru/genno/blob/0f577d2180bff440f3494daa6ac76af81939b66f/genno/core/exceptions.py#L66C1-L70C62) seems to be phrased differently, at least.

Reply:
Hi, yes. This exception is raised by the function .macro._validate_data()—in full it is something like `"Missing expected columns for gdp_calibration: {'year'}"`. pytest.raises(…, match=…) is treated as a regular expression, so the `.*` matches a bunch of characters in the middle.